src/lib/mina_net2/bitswap_block.ml

```diff
@@ -0,0 +1,237 @@
+    Bigstring.blit ~src:chunk ~src_pos:0 ~dst:block
+      ~dst_pos:(2 + (num_links * link_size))
+      ~len:chunk_size ;
+    let hash = Blake2.digest_bigstring block in
+    Hashtbl.set blocks ~key:hash ~data:block ;
+    Queue.enqueue link_queue hash
+  in
+  (* create the last block *)
+  create_block [] schema.last_block_data_size ;
+  if schema.total_blocks > 1 then (
+    (* create the data-only blocks *)
+    let num_data_only_blocks =
+      schema.total_blocks - schema.full_link_blocks - 1
+      - if schema.non_max_link_block_count > 0 then 1 else 0
+    in
+    for _ = 1 to num_data_only_blocks do
+      create_block [] max_data_chunk_size
+    done ;
+    (* create the non max link block, if there is one *)
+    if schema.non_max_link_block_count > 0 then (
+      let chunk_size =
+        max_block_size - 2 - (schema.non_max_link_block_count * link_size)
+      in
+      create_block (dequeue_links schema.non_max_link_block_count) chunk_size ;
+      (* create the max link blocks *)
+      let full_link_chunk_size =
+        max_block_size - 2 - (schema.max_links_per_block * link_size)
+      in
+      for _ = 1 to schema.full_link_blocks do
+        create_block
+          (dequeue_links schema.max_links_per_block)
+          full_link_chunk_size
+      done ) ) ;
+  assert (!remaining_data = 0) ;
+  assert (Queue.length link_queue = 1) ;
+  ( Blake2.Map.of_alist_exn (Hashtbl.to_alist blocks)
+  , Queue.dequeue_exn link_queue )
+
+let parse_block block =
+  if Bigstring.length block < 2 then Or_error.error_string "block too short"
+  else
+    let num_links = Bigstring.get_uint16_le block ~pos:0 in
+    if Bigstring.length block < 2 + (num_links * link_size) then
+      Or_error.error_string "block has invalid number of links"
+    else
+      let links =
+        List.init num_links ~f:(fun i ->
+            block
+            |> Bigstring.sub_shared ~pos:(2 + (i * link_size)) ~len:link_size
+            |> Bigstring.to_string |> Blake2.of_raw_string)
+      in
+      let data =
+        Bigstring.sub_shared block ~pos:(2 + (num_links * link_size))
+      in
+      Ok (links, data)
+
+let data_of_blocks blocks root_hash =
+  let open Or_error.Let_syntax in
+  let rec parse_chunks hash =
+    let%bind raw_block_data =
+      match Map.find blocks hash with
+      | Some data ->
+          return data
+      | None ->
+          Or_error.error_string "required block not found"
+    in
+    let%bind links, data = parse_block raw_block_data in
+    let%map tail = or_error_list_bind links ~f:parse_chunks in
+    data :: tail
+  in
+  let%map chunks = parse_chunks root_hash in
+  let total_data_size = List.sum (module Int) chunks ~f:Bigstring.length in
+  let data = Bigstring.create total_data_size in
+  ignore
+    ( List.fold_left chunks ~init:0 ~f:(fun i chunk ->
+          Bigstring.blit ~src:chunk ~src_pos:0 ~dst:data ~dst_pos:i
+            ~len:(Bigstring.length chunk) ;
+          i + Bigstring.length chunk)
```

Comment:
Nit: let's rename `i` to `dst_pos` so that it's not confused with simpler incremented counter